spec/models/database_spec.rb

```diff
@@ -0,0 +1,147 @@
+
+      model_class.reflect_on_all_associations(:belongs_to).each do |association|
+        migration = process_association(model_class, association)
+        migrations << migration unless migration.nil?
+      end
+    end
+
+    print_missing_foreign_key_warnings(migrations)
+
+    puts "The following models are marked as todo in #{__FILE__}:"
+    puts pending_models.join(", ")
+
+    migrations
+  end
+
+  def print_missing_foreign_key_warnings(migrations)
+    return if migrations.empty?
+
+    puts "Foreign key(s) appear to be absent from the database. " \
+         "You can add it/them using the following migration(s):"
+    puts migrations.join("\n")
+    puts "\nTo disable this warning, add the class name(s) of the model(s) to models_todo " \
+         "in #{__FILE__}"
+
+    return if ENV.fetch("OFN_WRITE_FOREIGN_KEY_MIGRATIONS", false)
+
+    puts "Migrations have not been written to disk. To write migrations to disk, please " \
+         "add OFN_WRITE_FOREIGN_KEY_MIGRATIONS=true to the file .env.test.local"
+  end
+
+  def process_association(model_class, association)
+    return if association.options[:polymorphic] || association.options[:optional]
+
+    foreign_key_table_name = determine_foreign_key_table_name(model_class, association)
+    foreign_key_column = association.options[:foreign_key] || "#{association.name}_id"
+    foreign_keys = model_class.connection.foreign_keys(model_class.table_name)
+
+    # Check if there is a foreign key that already exists for the column
+    return if foreign_keys.any? { |fk|
+                fk.column == foreign_key_column &&
+                fk.to_table == foreign_key_table_name
+              }
+
+    generate_migration(model_class, foreign_key_table_name, foreign_key_column)
+  end
+
+  def determine_foreign_key_table_name(model_class, association)
+    if association.options[:class_name]
+      class_name = association.options[:class_name].underscore.parameterize
+      foreign_key_table_name = class_name.tableize
+    else
+      foreign_key_table_name = association.class_name.underscore.parameterize.tableize
+      namespace = model_class.name.deconstantize
+
+      unless association.class_name.deconstantize == namespace || namespace == "" ||
+             ActiveRecord::Base.connection.table_exists?(foreign_key_table_name)
+        foreign_key_table_name = "#{namespace.underscore}_#{foreign_key_table_name}"
+      end
+    end
+
+    foreign_key_table_name
+  end
+
+  def generate_migration(model_class, foreign_key_table_name, foreign_key_column)
+    migration_name = "add_foreign_key_to_#{model_class.table_name}_" \
+                     "#{foreign_key_table_name}_#{foreign_key_column}"
+    migration_class_name = migration_name.camelize
+    millisecond_timestamp = Time.now.utc.strftime('%Y%m%d%H%M%S%L')
+    if millisecond_timestamp == @last_timestamp
+      millisecond_timestamp.to_i += 1
+    end
+    @last_timestamp = millisecond_timestamp
```

Comment:
This does not work:

```
millisecond_timestamp.to_i += 1
NoMethodError: undefined method `to_i=' for "20231004020203126":String
```

But even if `to_i` worked then you would create an integer and increment it but not update the original variable with the string.

```suggestion
    @last_creation_time ||= Time.new(0)
    creation_time = Time.now
    if creation_time <= @last_creation_time
      creation_time += 0.001.second
    end
    @last_creation_time = creation_time
    millisecond_timestamp = creation_time.utc.strftime('%Y%m%d%H%M%S%L')
```

I also used `<=` in case three migrations run in the same millisecond. Otherwise we get now, next millisecond and then now again. And that comparison doesn't work with nil, so I initialised the variable beforehand.

You can probably wrap that more easily in a method:
```rb
  # Generates a unique timestamp.
  #
  # We may create multiple migrations within the same second, maybe even millisecond.
  # So we add precision to milliseconds and increment on conflict.
  def generate_timestamp
    @last_creation_time ||= Time.new(0)

    creation_time = Time.now
    if creation_time <= @last_creation_time
      creation_time += 0.001.second
    end
    @last_creation_time = creation_time

    creation_time.utc.strftime('%Y%m%d%H%M%S%L')
  end
```

Reply:
Maybe not in this PR, but you could extract most of the logic into a lib class which can then be tested independently to cover these mistakes. And it could be moved into a gem to be used by other projects.

Reply:
@mkllnk OK, I extracted this out. Hopefully, this is ready to go.